colm/transpile/transpile.lm

```diff
@@ -0,0 +1,265 @@
+	exit( 1 )
+}
+
+bool time_constraint( Any: any )
+{
+	for UE: unary_expr in Any {
+		if match UE "time"
+			return true
+	}
+	return false
+}
+
+def time_range
+	Start: flux::expression
+	Stop: flux::expression
+[]
+
+void unquote_times( AE: ref<additive_expression> )
+{
+	for QI: quoted_identifier in AE {
+		Time: str = $QI
+		Time = prefix( suffix( Time, 1 ), Time.length - 2 )
+		QI.data = Time
+	}
+}
+
+bool check_time( Range: ref<time_range>, CE: comparison_expression )
+{
+	if match CE "time [Op: comparison_operator Rhs: additive_expression ]"
+			&& ( match Op "<" || match Op "<=" )
+	{
+		unquote_times( Rhs )
+		Range.Stop = parse flux::expression [Rhs]
+		return true
+	}
+
+	if match CE "time [Op: comparison_operator Rhs: additive_expression ]"
+			&& ( match Op ">" || match Op ">=" )
+	{
+		unquote_times( Rhs )
+		Range.Start = parse flux::expression [Rhs]
+		return true
+	}
+
+	return false
+}
+
+# Find time constraints and verify we can handle them.
+void yank_time( Range: ref<time_range>, Expr: ref<expr> )
+{
+	Error: bool = false
+	for CE: comparison_expression in Expr {
+		if ( time_constraint( CE ) ) {
+			if !match CE "( [expr] )" {
+				if !check_time( Range, CE ) {
+					send stderr "ERROR: wonky time constraint: [CE]
+					Error = true
+				}
+
+				CE = cons comparison_expression " __ERASE_ME__ "
+			}
+		}
+	}
+
+	if ( Error )
+		exit(1)
+}
+
+# Remove the expression components left behind by the time yanking.
+void clean_expr( Expr: ref<expr> )
+{
+	Modified: bool = true
+	while ( Modified ) {
+		Modified = false
+		for DE: disjunction_expression in Expr {
+			if match DE
+				"[ Left: disjunction_expression ] OR __ERASE_ME__"
+			{
+				DE = Left
+				Modified = true
+			}
+			else if match DE
+				"__ERASE_ME__ OR [ Right: conjunction_expression ]"
+			{
+				DE = cons disjunction_expression [Right]
+				Modified = true
+			}
+		}
+
+		for CE: conjunction_expression in Expr {
+			if match CE
+				"[ Left: conjunction_expression ] AND __ERASE_ME__"
+			{
+				CE = Left
+				Modified = true
+			}
+			else if match CE
+				"__ERASE_ME__ AND [ Right: comparison_expression ]"
+			{
+				CE = cons conjunction_expression [Right]
+				Modified = true
+			}
+		}
+
+		for CmpE: comparison_expression in Expr {
+			if match CmpE "( __ERASE_ME__ )" {
+				CmpE = cons comparison_expression "__ERASE_ME__"
+				Modified = true
+			}
+		}
+	}
+}
+
+void add_time_range( Output: output, Range: time_range )
+{
+	# Range stop.
+	cons Stop: flux::comma_property* []
+	if Range.Stop
+		Stop = parse flux::comma_property* ", stop: [$Range.Stop] "
+
+	send Output
+		"	|> range( start: [Range.Start] [Stop] )
+}
+
+void add_measurement_filter( Output: output, SS: select_stmt )
+{
+	# Add measurement as a filter.
+	for Measurement: measurement in SS.from_clause {
+		M: str = $Measurement
+		if match Measurement [unquoted_identifier]
+			M = "\"[M]\""
+		send Output
+			"	|> filter( fn: (r) => ( r._measurement == [M] ) )
```

Comment:
What would happen if you output syntactically incorrect flux? For example, what if you mis-typed the pipe forward operator:
```
send Output
    "    > filter( fn: (r) => r._measurement == [M] ) )
```
I guess it would be caught at runtime? Assuming `Output` is parsing the flux it receives?

Reply:
You got it! The output parser would fail at run-time and all subsequent text you send it is ignored. At the end you can retrieve the error message.

Using a parser object as the target of `send` statements is like writing to a file and spinning up a concurrent parsing process to read that file in as you write it. Only you can check the parser for success as you go. I don't actually do that here, just at the end.

If it succeeds you can also initiate the next transformation in the chain, in the same process, with access to all the same state, if you want.